### ISSUE TITLE
Bulk API

### ISSUE THREAD

**Issue by jorgenb** [babenkoivan/scout-elasticsearch-driver#37]
@babenkoivan I can not use the `/_bulk` endpoint for my setup but need to use either `/{index}/_bulk` or `{index}/{type}/_bulk`.

Is there an easy way to configure this driver to support this? It seems to default to `/_bulk`.

https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html

**Comment by babenkoivan** [comment]
Hey, have you tried to [configure](https://github.com/babenkoivan/scout-elasticsearch-driver#configuration) the bulk indexer? 

Just set the `index` option to `bulk` in `config/scout_elastic.php`. It has to use `/{index}/{type}/_bulk` url by default.

**Comment by jorgenb** [comment]
I have set the `index` option to `bulk`. Whenever I try to index it hits the endpoint `/_all/_bulk`.

**Comment by babenkoivan** [comment]
Hey, looked deeply into the code. You are right, currently only _all is used as endpoint. All the information about index and type is inside query body. It's made according this [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html):

```
POST _bulk
{ "index" : { "_index" : "test", "_type" : "type1", "_id" : "1" } }
{ "delete" : { "_index" : "test", "_type" : "type1", "_id" : "2" } }
```
So why isn't this appropriate for you?

**Comment by jorgenb** [comment]
@babenkoivan Thanks!

We use URL based security via a proxy and as such I can not use the `/_all` endpoint. Would you be able to produce a fix? I'll buy you a 🍻 :)

**Comment by babenkoivan** [comment]
It's better to ask guys responsible for [client](https://github.com/elastic/elasticsearch-php) to make request on `/_bulk` by default and not on `/_all/_bulk`.

But you can create a fix for you project by yourself as well. You can add in `scout_elastic.php` endpoint closure:

```
'client' => [
        'hosts' => [
            env('SCOUT_ELASTIC_HOST', 'localhost:9200')
        ],
        'endpoint' => function ($class) use ($transport, $serializer) {
            if ($class == 'Bulk') {
                return new MyBulkEndpoint($transport);
            } else {
                // Default handler
                $fullPath = '\\Elasticsearch\\Endpoints\\' . $class;
                if ($class === 'Msearch' || $class === 'MPercolate') {
                    return new $fullPath($transport, $serializer);
                } else {
                    return new $fullPath($transport);
                }
    }
}
]
```

`MyBulkEndpoint` must extend [Elasticsearch\Endpoints\Bulk](https://github.com/elastic/elasticsearch-php/blob/master/src/Elasticsearch/Endpoints/Bulk.php). Just override `getURI` method to `return '/_bulk'`

For more details have a look [here](https://www.elastic.co/guide/en/elasticsearch/client/php-api/current/_configuration.html#_set_the_endpoint_closure).

This is just a draft, an idea, I hope it will help you.

**Comment by jorgenb** [comment]
> It's better to ask guys responsible for client to make request to /_bulk by default rather than to /_all/_bulk.

Thanks, I'll look into it.

**Comment by jorgenb** [comment]
@babenkoivan 

Here is an example that forces the ES-PHP client to use `/{index}/{type}/_bulk`:

```
        $data = array();

        $elasticIndex = 'index';
        $elasticType = 'type';
        $client = ClientBuilder::create()->build();

        $params = [
            'index' => $elasticIndex,
            'body' => []
        ];

        $i = 1;
        foreach ($data as $body) {
            $params['body'][] = [
                'index' => [
                    '_index' => $elasticIndex,
                    '_type' => $elasticType,
                    '_id' => $i
                ]
            ];

            $params['body'][] = $body;

            // Every 1000 documents stop and send the bulk request
            if ($i % 1000 == 0) {
                $responses = $client->bulk($params);

                logger([$responses]);

                $this->info('Sending batch ...');

                // erase the old bulk request
                $params = [
                    'index' => $elasticIndex,
                    'body' => []
                ];

                // unset the bulk response when you are done to save memory
                unset($responses);
            }
            $i++;
        }

        // Send the last batch if it exists
        if (!empty($params['body'])) {
            $client->bulk($params);
        }
```

From the documentation @ https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html:

> The endpoints are /_bulk, /{index}/_bulk, and {index}/{type}/_bulk. When the index or the index/type are provided, they will be used by default on bulk items that don’t provide them explicitly.

Maybe it would be possible to specify some config option in your code that would set the `index` and the `type` when building the payload?

**Comment by babenkoivan** [comment]
Hey, I released v3.1.0 and there is some improvements regarding bulk indexing. Check if it works for you.

**Comment by jorgenb** [comment]
Seems to work just nicely for my setup. Thanks!
